docs/REST_API.md

```diff
@@ -0,0 +1,679 @@
+
+- Dataframe: expression matrix [nObs, nVar]
+  - Observations are cells
+  - Variables are genes (or equivalent variable, eg, isoforms)
+  - Observations and variables are identified by an matrix coordinate pair (index, base 0)
+  - Dataframe values must be defined (dense matrix)
+  - The underlying type of the dataframe is IEEE single precision float (IEEE 754 "decimal32" data type, float32 in many language environments)
+- Annotations (metadata):
+  - Observations/cells or variables/genes can be annotated
+  - Annotations are typed and described in a schema:
+    - Obs or var annotation
+    - Annotation name
+    - Annotation type (all encoding is JSON format):
+      - Float32 - scalar value, IEEE single precision float (decimal32), encoded as JSON number
+      - Int32 - scalar value, 2's complement signed Int 32 bit encoded as JSON number
+      - String - UTF-8 string, encoded as JSON string
+      - Boolean - boolean value, encoded as JSON true/false
+      - Categorical: value is an item from an _unordered_ enumeration. The enumeration schema is encoded as a JSON array. Enumeration categories may be a JSON number, string or boolean (or any combination thereof). Schema is encoded as a JSON array of enumeration category names.
+  - Annotation values will conform to their schema, _or_ be undefined.
+  - Annotation names are a string.
+  - There is one guaranteed annotation name on every dimension: "name". This should be used as the default display name, must be a string and must be unique for the dimension (ie, all observation names are unique and all variable names are unique).
+
+Missing from this model are ad hoc metadata/annotations not associated with a observation or variable - we can add that in the future if it is required (what ScanPy calls unstructured annotation).
+
+### Assumptions about the Server
+
+The following assumptions are made about server behavior:
+
+- The server will have a data schema for annotations
+- The server will have default algorithms for layout, clustering and differential expression (may add additional endpoints in the future)
+- The server will serve a default data set, including annotations, cluster assignment and 2D layout (put another way, the client can request this data without any specification about how it is generated).
+
+## Filters
+
+There are several common filters supported by various routes. Route descriptions supporting these filters will reference the filter types.
+
+### Annotation _Value_ Filters
+
+Filter by the value of specific observation or variable annotation values. For example, the `/data` routes support sub-selection by the value of the annotations, which might be used to filter by some specific metadata (eg, `tissue_type: lung`).
+
+Annotation value filters can be specified as a URL query parameter (GET), or a JSON object in a request body (POST)[^1].
```

Comment:
Can we replace POST with POST/PUT? Most of the urls are using PUT instead of POST so I think this is a bit confusing. There are a few instances of this.

Reply:
yes, will do